adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/X509AuthHandler.java

```diff
@@ -0,0 +1,96 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.adapter.http;
+
+import java.net.HttpURLConnection;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.Objects;
+
+import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.security.auth.x500.X500Principal;
+
+import org.eclipse.hono.client.HonoClient;
+import org.eclipse.hono.service.auth.device.HonoAuthHandler;
+import org.eclipse.hono.service.auth.device.HonoClientBasedAuthProvider;
+import org.eclipse.hono.util.CredentialsConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.impl.HttpStatusException;
+
+
+/**
+ * An authentication handler for client certificate based authentication.
+ *
+ */
+public class X509AuthHandler extends HonoAuthHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(X509AuthHandler.class);
+    private static final HttpStatusException UNAUTHORIZED = new HttpStatusException(HttpURLConnection.HTTP_UNAUTHORIZED);
+    private final HonoClient tenantServiceClient;
+
+    /**
+     * Creates a new handler for an authentication provider and a
+     * Tenant service client.
+     * 
+     * @param authProvider The authentication provider to use for verifying
+     *              the device identity.
+     * @param tenantServiceClient The client to use for determining the tenant
+     *              that the device belongs to.
+     */
+    public X509AuthHandler(final HonoClientBasedAuthProvider authProvider, final HonoClient tenantServiceClient) {
+        super(authProvider);
+        this.tenantServiceClient = Objects.requireNonNull(tenantServiceClient);
+    }
+
+    @Override
+    public void parseCredentials(final RoutingContext context, final Handler<AsyncResult<JsonObject>> handler) {
+
+        final Future<JsonObject> result = Future.future();
+        result.setHandler(handler);
+        if (context.request().isSSL()) {
+            try {
+                final Certificate[] clientChain = context.request().sslSession().getPeerCertificates();
+                if (clientChain[0] instanceof X509Certificate) {
+                    final X509Certificate clientCert = (X509Certificate) clientChain[0];
+                    tenantServiceClient.getOrCreateTenantClient().compose(tenantClient ->
+                        tenantClient.get(clientCert.getIssuerX500Principal())
+                    ).compose(tenant -> {
+                        final String subjectDn = clientCert.getSubjectX500Principal().getName(X500Principal.RFC2253);
```

Comment:
I think at this location it would make sense to add some way to allow people to extend this and extract more information from the certificate. I agree that this is fine for the default, but people like to encode all kind of information in certificates and IMHO it would make sense to call into some method like `extractInformationFromCertificate` which people may override if they need. And simply extract the subject for now.